natchez-http4s/src/main/scala/com/ovoenergy/effect/natchez/http4s/server/Configuration.scala

```diff
@@ -0,0 +1,198 @@
+package com.ovoenergy.effect.natchez.http4s.server
+
+import cats.data.Kleisli
+import cats.effect.Sync
+import cats.instances.map._
+import cats.kernel.Semigroup
+import cats.syntax.functor._
+import cats.syntax.semigroup._
+import cats.{Applicative, Apply}
+import com.ovoenergy.effect.natchez.http4s.server.Configuration.TagReader._
+import natchez.TraceValue
+import natchez.TraceValue.StringValue
+import org.http4s.util.{CaseInsensitiveString, StringWriter}
+import org.http4s.{Headers, Message, Request, Response}
+
+/**
+ * The tricky part about putting HTTP4s middleware into a library is that
+ * each user is likely to want to extract different tags from their requests
+ * according to whatever their organisation wide tagging policy is
+ *
+ * As such we define how to create tags from Requests / Responses with a `TagReader`
+ * which has a semigroup instance, allowing you to cherry pick what you want to extract
+ * and write your own extractors if required.
+ */
+case class Configuration[F[_]](
+  request: RequestReader[F],
+  response: ResponseReader[F]
+)
+
+object Configuration {
+
+  // just to shorten some lines
+  type Tags = Map[String, TraceValue]
+
+  /**
+   * A tag reader is essentially a function from an HTTP message to F[Tags]
+   * We need the effect because some tags may involve streaming the entity body
+   */
+  case class TagReader[F[_], -A](value: Kleisli[F, A, Tags]) extends AnyVal
+
+  object TagReader {
+
+    type MessageReader[F[_]] = TagReader[F, Message[F]]
+    type RequestReader[F[_]] = TagReader[F, Request[F]]
+    type ResponseReader[F[_]] = TagReader[F, Either[Throwable, Response[F]]]
+
+    /**
+     * Semigroup instance for TagReader
+     * so it is easy to read many tags from a request
+     */
+    implicit def sg[F[_]: Apply, A]: Semigroup[TagReader[F, A]] = {
+      implicit def applySg[B: Semigroup]: Semigroup[F[B]] = Apply.semigroup[F, B]
+      implicit val takeLast: Semigroup[TraceValue] = (_, b) => b
```

Comment:
Guessing this comes into play if two readers both generate tags with the same keys?

Reply:
indeed, hopefully that won't really ever happen